File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/UnitOfWorkEventPublisher.java
package com.ueueo.eventbus;

import com.ueueo.eventbus.distributed.IDistributedEventBus;
import com.ueueo.eventbus.local.ILocalEventBus;
import com.ueueo.uow.IUnitOfWorkEventPublisher;
import com.ueueo.uow.UnitOfWorkEventRecord;

import java.util.Collection;

public class UnitOfWorkEventPublisher implements IUnitOfWorkEventPublisher {
    private ILocalEventBus localEventBus;
    private IDistributedEventBus distributedEventBus;

    public UnitOfWorkEventPublisher(
            ILocalEventBus localEventBus,
            IDistributedEventBus distributedEventBus) {
        this.localEventBus = localEventBus;
        this.distributedEventBus = distributedEventBus;
    }

    @Override
    public void publishLocalEvents(Collection<UnitOfWorkEventRecord> localEvents) {
        for (UnitOfWorkEventRecord localEvent : localEvents) {
            localEventBus.publish(
                    localEvent.getEventType(),
                    localEvent.getGenericArgumentType(),
                    localEvent.getEventData(),
                    false
            );
        }
    }

    @Override
    public void publishDistributedEvents(Collection<UnitOfWorkEventRecord> distributedEvents) {
        for (UnitOfWorkEventRecord distributedEvent : distributedEvents) {
            distributedEventBus.publish(
                    distributedEvent.getEventType(),
                    distributedEvent.getGenericArgumentType(),
                    distributedEvent.getEventData(),
                    false
            );
        }
    }

}

File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/UnitOfWorkEventPublisher.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Volo.Abp.DependencyInjection;
using Volo.Abp.EventBus.Distributed;
using Volo.Abp.EventBus.Local;
using Volo.Abp.Uow;

namespace Volo.Abp.EventBus;

[Dependency(ReplaceServices = true)]
public class UnitOfWorkEventPublisher : IUnitOfWorkEventPublisher, ITransientDependency
{
    private readonly ILocalEventBus _localEventBus;
    private readonly IDistributedEventBus _distributedEventBus;

    public UnitOfWorkEventPublisher(
        ILocalEventBus localEventBus,
        IDistributedEventBus distributedEventBus)
    {
        _localEventBus = localEventBus;
        _distributedEventBus = distributedEventBus;
    }

    public void PublishLocalEventsAsync(IEnumerable<UnitOfWorkEventRecord> localEvents)
    {
        for (var localEvent in localEvents)
        {
            _localEventBus.PublishAsync(
                localEvent.EventType,
                localEvent.EventData,
                onUnitOfWorkComplete: false
            );
        }
    }

    public void PublishDistributedEventsAsync(IEnumerable<UnitOfWorkEventRecord> distributedEvents)
    {
        for (var distributedEvent in distributedEvents)
        {
            _distributedEventBus.PublishAsync(
                distributedEvent.EventType,
                distributedEvent.EventData,
                onUnitOfWorkComplete: false,
                useOutbox: distributedEvent.UseOutbox
            );
        }
    }
}
